lib/inspec/rule.rb

```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
       @__checks.push([describe_or_expect, values, block])
     end
 
+    # Look for an input with a matching ID, and if found, apply waiver
+    # skipping logic. Basically, if we have a current waiver, and it says
+    # to skip, we'll replace all the checks with a dummy check (same as
+    # only_if mechanism)
+    # Double underscore: not intended to be called as part of the DSL
+    def __apply_waivers
+      input_name = "waiver_#{@__rule_id}" # TODO: control ID slugging
+      registry = Inspec::InputRegistry.instance
+      input = registry.inputs_by_profile[@__profile_id].dig(input_name)
+      return unless input
+
+      input = input.value
+
+      # Waivers should have a hash value with keys including skip and
+      # expiration_date. We only care here if it has a skip key and it
+      # is yes-like, since all non-skipped waiver operations are handled
+      # during reporting phase.
+      return unless input.key?("skip")
+      return unless input["skip"].to_s.match(/y|yes|true/i)
+
+      # OK, the intent is to skip. Does it have an expiration date, and
+      # if so, is it in the future?
+      expiry = input["expiration_date"]
+      if expiry
+        if expiry.is_a?(Date)
+          # It appears that yaml.rb automagically parses dates for us
+          return if expiry < Date.today # If the waiver expired, return - no skip applied
+        elsif expiry.match(/never/i) # any other values?
+          # Do nothing, fall through
+        else
+          raise Inspec::Exceptions::ResourceFailed, "Unable to parse waiver expiration date '#{expiry}' for control #{@__rule_id}"
+        end
+      end
+
+      # OK, apply a skip.
+      @__skip_rule[:result] = true
+      @__skip_rule[:message] = "Skipped due to waiver: #{input["justification"] || "(no reason given)"}"
```

Comment:
I might be misunderstanding the point of `@__skip_rule`, but to me this should be passing through the expiration date and whether it was skipped?

...or does the backend need to re-process all of the same input/waiver data?

The answer to this question determines the overall implementation of step 3.

Reply:
Also, since the tests are laid out for step 3... should @miah be working _on top_ of your branch?

Reply:
As I understand it, `@__skip_rule` is a data structure that determines if the control has been marked to be skipped. If `@__skip_rule[:result]` is true, then the runtime will "skip" the tests (by replacing all of them with one dummy `os` check, see `prepare_checks`).  `only_if` uses the same mechanism.

I'm all for altering the fields in that structure to make it more flexible - right now it adds bad messaging, assuming all skips are "Only if." We could also invent a new mechanism.

All of this here, though, only affects the textual messaging associated with the control. For step three, we're wanting new separate fields in the JSON output. And I think - with some code deduplication - it should be the same sort of logic we see in step 2 - read from the Input, then consider if it was expired, was it to be skipped, etc.  I'm not sure how much of the Rule datastructure we have access to during the reporting phase (if any) - clear that's something to look into ASAP.

Reply:
OK, I've located where we'd stash such data in the RSpec example metadata - we have an opportunity here - https://github.com/inspec/inspec/blob/d1ffce727828853222b80ae018d969b2f6c2bc74/lib/inspec/runner_rspec.rb#L166 - however data structure there does not propagate to the run_data hash passed to the reporter structure. I'm still looking at that.

Reply:
I found where the RSpec metadata may be propagated to the reporter Hash data structure - here - https://github.com/inspec/inspec/blob/master/lib/inspec/formatters/base.rb#L152 .

At this point we could concentrate all waiver logic in step 2, and only do display in step 3.

Reply:
I've cleaned up the way that the messaging is handled, and placed all of the business logic around waivers in step 2. AFAIK in step 3 now we just need to display fields - I've updated the design doc with notes on what should be available in the `run_data` structure.